### PR TITLE
Switch build system to CMake

### PR DESCRIPTION
**Pull request by kkaefer**
revert #28

We want to switch to CMake from gyp because:

* gyp isn't maintained
* Android Studio has explicit CMake integration
* Qt Creator has explicit CMake integration
* CMake automatically rebuilds the project if CMake files changed
* Less complicated build system, fewer bash files

### CODE REVIEW

**Comment by jfirebaugh** [comment]
> revert #28

Hahah, wow.

**Comment by brunoabinader** [comment]
screen shot 2016-06-15 at 9 23 45 am = ❤️

**Comment by tmpsantos** [comment]
Kudos.

**Comment by tmpsantos** [comment]
Made progress with Qt today. All the configurations, tests, etc work just fine on Linux.

https://github.com/mapbox/mapbox-gl-native/tree/tmpsantos-cmake

**Comment by ivovandongen** [comment]
Got most of the android build sorted out on: https://github.com/mapbox/mapbox-gl-native/tree/ivovandongen-cmake

Links and runs on arm-v5 and arm-v7, but results in linker errors on others still:

```
[100%] Linking CXX shared library libmapbox-gl.so
/Users/ivo/git/mapbox-gl-native/mason_packages/osx-x86_64/android-ndk/arm64-21-r12b/bin/../lib/gcc/aarch64-linux-android/4.9.x/../../../../aarch64-linux-android/bin/ld: skipping incompatible /Users/ivo/git/mapbox-gl-native/mason_packages/android-arm-v7-9/zlib/system/lib/libz.so when searching for -lz
../../../mason_packages/android-arm-v7-9/geojson/0.1.4/lib/libgeojson.a: error adding symbols: File in wrong format
```

**Comment by ivovandongen** [comment]
All ABIs are building now. Also added the copy targets. The libraries are insanely big though, 122MB instead of ~13MB. 

After reading `mason.cmake` I have a better idea how to get rid of `toolchain.sh`, if we want.

**Comment by ivovandongen** [comment]
@kkaefer I've squashed the  android part into one commit. Need to test it a bit to be sure there are no regressions on Monday. @tobrun Mind taking a look ? Branch is [here](https://github.com/mapbox/mapbox-gl-native/tree/ivovandongen-cmake).

**Comment by tobrun** [comment]
I have been testing with `make apackage -j4`and verified that following abi were build:
 - `arm-v5`  (armeabi configuration)
 - `arm-v7`
 - `arm-v8` (64 bit arm)
 - `x86`
 - `x86-64`
 - `mips`

I did some runs with this on multi devices as:
 - arm-v7 emulator
 - genymotion virtual x86 device
 - arm-v7 device 

I also tested out backwards compatibility issues with for arm-v7 on armabi (v5-v6):
- `make android-lib-arm-v5` on arm-v7 emulator
- `make android-lib-arm-v7` on arm-v7 emulator

Everything is looking 👍 @ivovandongen, can't wait to get this merged and make this part of our IDE workflow in https://github.com/mapbox/mapbox-gl-native/issues/1753.

**Comment by ivovandongen** [comment]
@tobrun Thanks for the extensive testing! I would also love to have this in one IDE eventually. Need to check out the new experimental gradle plugin. Seems like that is going to be preferred over the "classic" Android.mk.

**Comment by ivovandongen** [comment]
@kkaefer @tmpsantos Changes from #5656 got lost in translation, added them back in. I've added them in the android `toolchain.sh` file, where it was in the general `mbgl.gyp` before. If this is not the preferred place let me know.

Also, could either of you go over my changes to see if I haven't missed other subtle settings?

File: .gitignore [hunk 1]
```diff
@@ -25,3 +25,4 @@ xcuserdata
 /platform/ios/benchmark/assets/tiles/mapbox.mapbox-terrain-v2,mapbox.mapbox-streets-v6
 **/token
 /platform/macos/macos.xcworkspace/xcshareddata/macos.xcscmblueprint
+/platform/ios/pkg
```
Comment (1ec5) [inline .gitignore]: What about /platform/macos/pkg? It doesn’t look like the `OUTPUT` variable in platform/macos/scripts/package.sh has been changed from `build/macos/pkg` to `platform/macos/pkg` as for iOS.

File: Makefile [hunk 2]
```diff
@@ -367,6 +367,45 @@ test-node: node
...
+
+build/android-$1/$(BUILDTYPE)/Makefile: build/android-$1/$(BUILDTYPE)/toolchain.cmake platform/android/config.cmake
+	cd build/android-$1/$(BUILDTYPE) && cmake ../../.. \
+		-DCMAKE_TOOLCHAIN_FILE=build/android-$1/$(BUILDTYPE)/toolchain.cmake \
+		-DCMAKE_BUILD_TYPE=$(BUILDTYPE) \
+		-DCMAKE_EXPORT_COMPILE_COMMANDS=ON \
+		-DMBGL_PLATFORM=android
+
+android-lib-$1: build/android-$1/$(BUILDTYPE)/Makefile
+	$(MAKE) -j$(JOBS) -C build/android-$1/$(BUILDTYPE) all
```
Comment (tmpsantos) [inline Makefile]: We could use $(NINJA) here by passing `-G Ninja` to the `cmake` command.

Comment (kkaefer) [inline Makefile]: Android Studio 2.2 also uses a Ninja derivative (`-G"Android Gradle - Ninja"`)

**Comment by tmpsantos** [comment]
Readiness checklist (compiles, runs, tests pass):

[ ] GLFW Linux
[ ] GLFW OSX
[ ] iOS
[ ] macOS
[ ] Android armeabi
[ ] Android armv7
[ ] Android aarch64
[ ] Android x86
[ ] Android x86_64
[ ] Qt4 Linux
[ ] Qt4 OSX
[ ] Qt5 Linux
[ ] Qt5 OSX

**Comment by mikemorris** [comment]
On node/macOS, it looks like this breaks the shim I had set up to pass through `BUILDTYPE=Debug`

**Comment by mikemorris** [comment]
Found the problem 😄 

[`"install": "node-pre-gyp install --fallback-to-build=false || BUILDTYPE=Release make node",`](https://github.com/mapbox/mapbox-gl-native/blob/9ae2fc954d6f621ac3c92d1fc98e276ac0f26d29/package.json#L35)

**Comment by tmpsantos** [comment]
Android build completes on Linux but app fails to launch due to missing symbols on the shared library. Many new files were added recently that are missing.

File: platform/ios/scripts/package.sh [hunk 3]
```diff
@@ -5,9 +5,9 @@ set -o pipefail
 set -u
 
 NAME=Mapbox
-OUTPUT=build/ios/pkg
+OUTPUT=platform/ios/pkg
```
Comment (1ec5) [inline platform/ios/scripts/package.sh]: @kkaefer, were you planning to change this back to build/ios/pkg/, or were you planning to change [this line](https://github.com/mapbox/mapbox-gl-native/pull/5359/commits/480869cfcad7368b95f17c9f59e8ef008053fe1b#diff-4b90a74e5e02a5a09ced6c70a733a0feR8) to match?

**Comment by friedbunny** [comment]
Several of the packaging commands for the iOS SDK are broken, e.g., `make ipackage`, `make iframework`, `make iframework-strip`:

```
▸ Compiling MGLMapView.mm
** BUILD FAILED **


The following build commands failed:
	Libtool /Users/jason/Documents/Mapbox/mapbox-gl-native/build/ios/ios.build/Debug-iphonesimulator/static.build/Objects-normal/i386/libMapbox.a normal i386
	Libtool /Users/jason/Documents/Mapbox/mapbox-gl-native/build/ios/ios.build/Debug-iphonesimulator/static.build/Objects-normal/x86_64/libMapbox.a normal x86_64
(2 failures)
▸ Compiling MGLMapCamera.mm
▸ Compiling MGLAnnotationView.mm
▸ Compiling MGLUserLocationAnnotationView.m
▸ Compiling MGLMapboxEvents.m
▸ Compiling MGLPolygon.mm
▸ Compiling NSProcessInfo+MGLAdditions.m
▸ Compiling MGLOfflineStorage.mm
▸ Compiling MGLCompactCalloutView.m
▸ Compiling MGLTilePyramidOfflineRegion.mm
▸ Compiling MGLAccountManager.m
▸ Compiling MGLPolyline.mm
▸ Building library libMapbox.a

❌  error: /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/libtool: can't locate file for: -lmbgl-platform-ios

❌  error: /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/libtool: file: -lmbgl-platform-ios is not an object file (not allowed in a library)
```

**Comment by ivovandongen** [comment]
@tmpsantos re:android I've added a bunch more today on master. I can sort that out tmr if needed.

**Comment by kkaefer** [comment]
@friedbunny fixed packaging for iOS
@ivovandongen added the missing Android files
@tmpsantos fixed Qt4 build on macOS

**Comment by tmpsantos** [comment]
Building armv7 works on Linux, but when building armv8 on Linux, I get paths to armv7 from mason (even after a clean build):

```
./../../mason_packages/android-arm-v7-9/geojson/0.1.4/lib/libgeojson.a ../../../mason_packages/android-arm-v7-9/sqlite/3.9.1/lib/libsqlite3.a ../../../mason_packages/android-arm-v7-9/nunicode/1.6/lib/libnu.a ../../../mason_packages/android-arm-v7-9/libpng/1.6.20/lib/libpng.a -lz -lm ../../../mason_packages/android-arm-v7-9/libjpeg-turbo/1.4.2/lib/libjpeg.a ../../../mason_packages/android-arm-v7-9/libzip/0.11.2/lib/libzip.a -lz -llog -landroid -lEGL -lGLESv2 -lstdc++ -latomic && :
../../../mason_packages/android-arm-v7-9/geojson/0.1.4/lib/libgeojson.a: error adding symbols: File in wrong format
clang38++: error: linker command failed with exit code 1 (use -v to see invocation)
```

**Comment by friedbunny** [comment]
iOS changes look good — dynamic and static frameworks build and run in release and debug, all of the symbols are where they should be, and bitcode is still enabled. (I’ve become quite paranoid about these things.)

Recompilation is also much faster than with gyp, which made it easier to verify these changes. 🙇

**Comment by tmpsantos** [comment]
Android + Linux worked fine for me now on 6034ac7.

**Comment by incanus** [comment]
Docs need updated. 

```bash
$ make iproj
mkdir -p build/ios
(cd build/ios && cmake -G Xcode ../.. \
		-DCMAKE_TOOLCHAIN_FILE=../../platform/ios/toolchain.cmake \
		-DMBGL_PLATFORM=ios)
/bin/sh: cmake: command not found
make: *** [build/ios/mbgl.xcodeproj] Error 127
```

**Comment by bleege** [comment]
I just updated the install docs to list CMake for all binding flavors.

**Comment by nitrag** [comment]
@bleege Where are we supposed to install CMake? I'm getting this error:

```
Failed to execute '/usr/local/bin/node /usr/local/lib/node_modules/npm/node_modules/node-gyp/bin/node-gyp.js configure --fallback-to-build=false --module=/Users/ryan/Documents/github/mapbox-gl-native/lib/mapbox-gl-native.node --module_name=mapbox-gl-native --module_path=/Users/ryan/Documents/github/mapbox-gl-native/lib' (1)
mkdir -p build/macos
(cd build/macos && cmake -G Xcode ../..)
Re-run cmake no build system arguments
-- The CXX compiler identification is AppleClang 8.0.0.8000038
-- The C compiler identification is AppleClang 8.0.0.8000038
-- Check for working CXX compiler using: Xcode
-- Check for working CXX compiler using: Xcode -- works
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- Check for working C compiler using: Xcode
-- Check for working C compiler using: Xcode -- works
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Detecting C compile features
-- Detecting C compile features - done
CMake Error at cmake/mbgl.cmake:16 (message):
  Can't find node-cmake
Call Stack (most recent call first):
  CMakeLists.txt:4 (include)


-- Configuring incomplete, errors occurred!
See also "/Users/ryan/Documents/github/mapbox-gl-native/build/macos/CMakeFiles/CMakeOutput.log".
make: *** [build/macos/mbgl.xcodeproj] Error 1
```

I've tried install globally too.

**Pull request merged**